examples/src/main/scala/basic/GPT2.scala

```diff
@@ -1,378 +1,405 @@
-//         println("Successfully loaded WPE parameters")
-//         val wte = load2("wte.weight", Axis[Vocab], Axis[Embedding])
-//         println("Successfully loaded WTE parameters")
-//         val ln_f = loadLN("ln_f")
-//         println("Successfully loaded final LayerNorm parameters")
-
-//         val layers = (0 until 12).map { i =>
-//             val prefix = s"h.$i"
-//             val ln1 = loadLN(s"$prefix.ln_1")
-//             val ln2 = loadLN(s"$prefix.ln_2")
-
-//             val cAttn = loadLinear(s"$prefix.attn.c_attn", Axis[Embedding], Axis[QKV])
-//             val cProj = loadLinear(s"$prefix.attn.c_proj", Axis[Heads |*| Value], Axis[Embedding])
-//             val attn = MultiHeadAttentionParams(cAttn, cProj, numHeads = 12)
-
-//             val c_fc = loadLinear(s"$prefix.mlp.c_fc", Axis[Embedding], Axis[Inner])
-//             val c_proj = loadLinear(s"$prefix.mlp.c_proj", Axis[Inner], Axis[Embedding])
-//             val mlp = MLPParams(c_fc, c_proj)
-//             println(s"Successfully loaded layer $i parameters")
-
-//             HiddenParams(ln1, attn, ln2, mlp)
-//         }.toList
-//         println("Successfully loaded all layers parameters")
-
-//         val params = GPT2Params(wpe, wte, layers, ln_f)
-//         val gpt2 = GPT2(params)
-//         val inference = Inference(gpt2, Tokenizer(tiktoken.get_encoding("gpt2")))
-//         // val stream = inference("Hello, my name is")
-//         // stream.foreach(println)
+package examples.basic
+
+import dimwit.*
+import dimwit.Conversions.given
+
+import nn.ActivationFunctions.*
+
+// Dimensions
+trait Vocab derives Label // 50257
+trait Embedding derives Label // 768
+trait Context derives Label // 1024
+trait EmbeddingMixer derives Label // 3072
+
+trait Batch derives Label
+
+case class LayerNormalizationParams(
+    weight: Tensor1[Embedding, Float],
+    bias: Tensor1[Embedding, Float]
+)
+
+case class LinearLayerParams[In, Out](
+    weight: Tensor2[In, Out, Float],
+    bias: Tensor1[Out, Float]
+)
+
+case class ProjectionLayerParams[In, Out](
+    weight: Tensor2[In, Out, Float]
+)
+
+// Values stored as Query, Key, Value triplets
+// type QKV[L <: String] = L * L * L
+// case class AttentionParams(
+// Fused Q,K,V projection (768 -> 2304)
+//    cAttn: LinearLayerParams[Embedding, QKV[Embedding]],
+// Output projection (768 -> 768)
+//    cProj: LinearLayerParams[Embedding, Embedding]
+//)
+
+trait Heads derives Label
+trait Key derives Label
+trait Query derives Label
+trait Value derives Label
+trait HeadValue derives Label
+
+case class HeadsParams[Kind](val weights: Tensor3[Heads, Embedding, Kind, Float], val bias: Tensor2[Heads, Kind, Float])
+
+case class MultiHeadAttentionParams(
+    wq: HeadsParams[Query],
+    wk: HeadsParams[Key],
+    wv: HeadsParams[HeadValue],
+    proj: LinearLayerParams[Heads |*| HeadValue, Embedding]
+) derives ToPyTree
+
+type QKV = Heads |*| Query |*| Heads |*| Key |*| Heads |*| HeadValue
```

Comment:
This part is tricky. A comment would be good explaining the splitting (motivation, types and logic)